saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/Saml2AuthenticationRequestContext.java

```diff
@@ -0,0 +1,235 @@
+import org.springframework.util.Assert;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+
+import static org.springframework.util.StringUtils.hasText;
+
+/**
+ * Data holder for information required to create an {@code AuthNRequest}
+ * to be sent from the service provider to the identity provider
+ * https://www.oasis-open.org/committees/download.php/35711/sstc-saml-core-errata-2.0-wd-06-diff.pdf (line 2031)
+ *
+ * @see {@link Saml2AuthenticationRequestFactory}
+ * @since 5.3
+ */
+public final class Saml2AuthenticationRequestContext {
+	private final RelyingPartyRegistration relyingPartyRegistration;
+	private final String issuer;
+	private final String assertionConsumerServiceUrl;
+	private final String relayState;
+
+	private Saml2AuthenticationRequestContext(
+			RelyingPartyRegistration relyingPartyRegistration,
+			String issuer,
+			String assertionConsumerServiceUrl,
+			String relayState) {
+		Assert.hasText(issuer, "issuer cannot be null or empty");
+		Assert.notNull(relyingPartyRegistration, "relyingPartyRegistration cannot be null");
+		Assert.hasText(assertionConsumerServiceUrl, "spAssertionConsumerServiceUrl cannot be null or empty");
+		this.issuer = issuer;
+		this.relyingPartyRegistration = relyingPartyRegistration;
+		this.assertionConsumerServiceUrl = assertionConsumerServiceUrl;
+		this.relayState = relayState;
+	}
+
+
+	/**
+	 * returns the issuer, the local SP entity ID, for this authentication request.
+	 * This property should be used to populate the {@code AuthNRequest.Issuer} XML element.
+	 * This value typically is a URI, but can be an arbitrary string.
+	 * @return issuer
+	 */
+	public String getIssuer() {
+		return this.issuer;
+	}
+
+	/**
+	 * returns the destination, the WEB Single Sign On URI, for this authentication request.
+	 * This property populates the {@code AuthNRequest#Destination} XML attribute.
+	 * @return destination
+	 */
+	public String getDestination() {
+		return this.getRelyingPartyRegistration().getIdpWebSsoUrl();
+	}
+
+	/**
+	 * Returns the {@link RelyingPartyRegistration} configuration for which the AuthNRequest is intended for.
+	 * @return the {@link RelyingPartyRegistration} configuration
+	 */
+	public RelyingPartyRegistration getRelyingPartyRegistration() {
+		return this.relyingPartyRegistration;
+	}
+
+	/**
+	 * Returns the desired {@code AssertionConsumerServiceUrl} that this SP wishes to receive the
+	 * assertion on. The IDP may or may not honor this request.
+	 * This property populates the {@code AuthNRequest#AssertionConsumerServiceURL} XML attribute.
+	 * @return the AssertionConsumerServiceURL value
+	 */
+	public String getAssertionConsumerServiceUrl() {
+		return assertionConsumerServiceUrl;
+	}
+
+	/**
+	 * Returns a list of credentials that can be used to sign the {@code AuthNRequest} object
+	 * @return signing credentials
+	 */
+	public List<Saml2X509Credential> getCredentials() {
```

Comment:
I'd probably leave this out. It is a nice convenience method, but then what's the argument against also adding in `getSigningCredentials()` from `RelyingPartyRegistration`? Adding this convenience method might be premature, and in one implementation it'd probably be better to call `getRelyingPartyRegistration().getSigningCredentials()` anyway, wouldn't it?

Reply:
Completed